server/src/main/java/io/pravega/schemaregistry/storage/client/TableStore.java

```diff
@@ -0,0 +1,392 @@
+import io.pravega.controller.store.stream.PravegaTablesStoreHelper;
+import io.pravega.controller.store.stream.StoreException;
+import io.pravega.controller.util.RetryHelper;
+import io.pravega.schemaregistry.server.rest.ServiceConfig;
+import io.pravega.schemaregistry.storage.StoreExceptions;
+import lombok.Data;
+import lombok.NonNull;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.AbstractMap;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static io.pravega.controller.server.WireCommandFailedException.Reason.ConnectionDropped;
+import static io.pravega.controller.server.WireCommandFailedException.Reason.ConnectionFailed;
+
+/**
+ * Wrapper class over {@link PravegaTablesStoreHelper}. Its implementation abstracts the caller classes from the library
+ * used for interacting with pravega tables.
+ */
+@Slf4j
+public class TableStore {
+    public static final String SCHEMA_REGISTRY_SCOPE = "_schemaregistry";
+    private static final int NUM_OF_RETRIES = 15; // approximately 1 minute worth of retries
+    private final SegmentHelper segmentHelper;
+    private final HostStoreImpl hostStore;
+    private final int numOfRetries;
+    private final ScheduledExecutorService executor;
+    private final Cache<TableCacheKey, Version.VersionedRecord<?>> cache;
+    private final Function<String, String> tokenSupplier;
+    private final CompletableFuture<Void> createScope;
+    private final Cache<String, String> tokenCache;
+    
+    public TableStore(ClientConfig clientConfig, ServiceConfig serviceConfig, ScheduledExecutorService executor) {
+        ConnectionFactoryImpl connectionFactory = new ConnectionFactoryImpl(clientConfig);
+        hostStore = new HostStoreImpl(clientConfig, executor);
+        segmentHelper = new SegmentHelper(connectionFactory, hostStore);
+        this.executor = executor;
+        this.tokenSupplier = x -> {
+            String[] splits = x.split("/");
+            return hostStore.getController().getOrRefreshDelegationTokenFor(splits[0], splits[1]).join();
+        };
+        numOfRetries = NUM_OF_RETRIES;
+        this.cache = CacheBuilder.newBuilder()
+                                 .maximumSize(10000)
+                                 .build();
+
+        tokenCache = CacheBuilder.newBuilder()
+                    .maximumSize(10000)
+                    .build();
+        this.createScope = createScope(); 
+    }
+    
+    public CompletableFuture<Void> createScope() {
+        return Retry.indefinitelyWithExpBackoff("Failed to create scope. Retrying...")
```

Comment:
Do we expect the exception to be always transient? Is that a fair assumption? I'd think no.

Reply:
We cannot do anything until we have created SchemaRegistry scope in pravega as we create all the tables under that scope.